dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NormalizeLoops.java

```diff
@@ -69,5 +104,46 @@ public void run(Function function) {
 
             counter++;
         }
+
+        // Guarantees header is the only entry point
+        for (Label label : function.getEvents(Label.class)) {
+            final List<CondJump> backJumps = label.getJumpSet().stream()
+                    .filter(j -> j.getLocalId() > label.getLocalId())
+                    .sorted()
+                    .toList();
+            final Label loopBegin = label;
+            final CondJump uniqueBackJump = backJumps.get(0);
+
+            final List<Label> loopBodyLabels = loopBegin.getSuccessor().getSuccessors().stream()
+                    .takeWhile(ev -> ev != uniqueBackJump)
+                    .filter(Label.class::isInstance).map(Label.class::cast)
+                    .toList();
+            for (Label l : loopBodyLabels) {
+                final List<CondJump> externalEntries = l.getJumpSet().stream()
+                        .filter(j -> j.getLocalId() < loopBegin.getLocalId() ||
+                                j.getLocalId() > uniqueBackJump.getLocalId())
+                        .toList();
+
+                for (CondJump fromOutside : externalEntries) {
+
+                    final Register jumpedFromOutside = function.newRegister("__jumpedFromOutside",
+                            types.getBooleanType());
+                    final Local setJumpedFromOutside = EventFactory.newLocal(jumpedFromOutside, expressions.makeTrue());
+                    final CondJump jumpToHeader = EventFactory.newGoto(loopBegin);
+                    final Label internalLabel = fromOutside.getLabel();
```

Comment:
You need to consider jumps to the header though because you still need to set the value of the entry point variable to 0.
```
goto LoopHeader
...
LoopHeader:
...
goto LoopHeader
```
This code needs to be transformed to
```
__loopEntryPointId = 0;
goto LoopHeader;
...
__loopEntryPointId = 0; // This gets skipped by any direct jump to the loop header!
LoopHeader:
...
__loopEntryPointId = 0; // The backjumps needs this to reset the entry point.
goto LoopHeader;
```

There are different solutions though. For example, all outside jumps that go directly to the loop header, could be retargeted to a preheader of the loop (the preheader needs to get generated of course) that initializes the variable as desired.
The backjump could also avoid resetting the variable if you do one of the following:
- reset the variable in each special entry (probably leads to harder-to-analyze/optimize code)
- reset the variable on loop entry. In this case the value needs to get stored in a second helper variable right before resetting, so that the value is still available for the switch-case.

Reply:
My current solution (see latest commit a few minutes ago) should already take care of this